web/src/tracking/withActivation.tsx

```diff
@@ -0,0 +1,271 @@
+    if (update.FoundReferences) {
+        logUserEvent(GQL.UserEvent.CODEINTELREFS)
+    }
+}
+
+interface Props {
+    authenticatedUser: GQL.IUser | null
+}
+
+interface State {
+    completed?: ActivationCompletionStatus
+}
+
+/**
+ * Modifies the input component to return a component that includes the activation status in the
+ * `activation` field of its props.
+ */
+export const withActivation = <P extends ActivationProps & Props>(Component: React.ComponentType<P>) =>
+    class WithActivation extends React.Component<Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>, State> {
+        private subscriptions = new Subscription()
+        private componentUpdates = new Subject<Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>>()
+        public state: State = {}
+
+        /**
+         * Calling `next` triggers refetches. This ensures at most one refetch request is outstanding
+         * at any given time.
+         */
+        private refetches = new Subject<void>()
+
+        private updates = new Subject<Partial<ActivationCompletionStatus>>()
+
+        public componentDidMount(): void {
+            const authenticatedUser: Observable<GQL.IUser | null> = this.componentUpdates.pipe(
+                startWith(this.props),
+                map(props => props.authenticatedUser),
+                distinctUntilChanged()
+            )
+            const serverCompletionStatus: Observable<ActivationCompletionStatus> = combineLatest(
+                authenticatedUser,
+                this.refetches
+            ).pipe(
+                switchMap(([authenticatedUser]) =>
+                    authenticatedUser ? fetchActivationStatus(authenticatedUser.siteAdmin) : []
+                )
+            )
+            const localCompletionStatus: Observable<Partial<ActivationCompletionStatus>> = merge(
+                authenticatedUser.pipe(map(() => null)), // reset on new authenticated user
+                this.updates
+            ).pipe(
+                tap(update => update && recordUpdate(update)),
+                scan((prev, next) => (next ? { ...prev, ...next } : {})),
+                map(p => p || {})
+            )
+            this.subscriptions.add(
+                combineLatest(serverCompletionStatus, localCompletionStatus)
+                    .pipe(
+                        map(([serverCompletionStatus, localCompletionStatus]) => ({
+                            ...serverCompletionStatus,
+                            ...localCompletionStatus,
+                        }))
+                    )
+                    .subscribe(completed => this.setState({ completed }))
+            )
+        }
+
+        public componentWillUnmount(): void {
+            this.subscriptions.unsubscribe()
+        }
+
+        public componentDidUpdate(): void {
+            this.componentUpdates.next(this.props as Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>)
```

Comment:
I think you can get rid of this cast if you make the Subject readonly (see above)
```suggestion
            this.componentUpdates.next(this.props)
```